bindings/node/lib/clientEncryption.js

```diff
@@ -549,6 +554,75 @@ module.exports = function (modules) {
       return value;
     }
 
+    /**
+     * Creates a collection that has encrypted document fields
+     *
+     * @template {TSchema} - Schema for the collection being created
+     * @param {Db} db - A database to create the collection in
+     * @param {string} name - The name of the collection to be created
+     * @param {object} options - Options for createDataKey and for createCollection.
+     * @param {string} options.provider - provider name
+     * @param {ClientEncryptionCreateDataKeyProviderOptions} options.createDataKeyOptions - options to pass to createDataKey
+     * @param {CreateCollectionOptions} options.createCollectionOptions - options to pass to createCollection, must include `encryptedFields`
+     * @returns {Promise<{ collection: Collection<TSchema>, encryptedFields: Document }>} - created collection and generated encryptedFields
+     * @throws {MongoCryptCreateEncryptedCollectionError} - If part way through the process createDataKey fails, an error will be rejected that has the `encryptedFields` that were created.
+     */
+    async createEncryptedCollection(db, name, options) {
+      const {
+        provider,
+        createDataKeyOptions = {},
+        createCollectionOptions: {
+          encryptedFields: { ...encryptedFields },
+          ...createCollectionOptions
+        }
+      } = options;
+
+      if (Array.isArray(encryptedFields.fields)) {
+        encryptedFields.fields = [...encryptedFields.fields];
+
+        const createDataKeyPromises = [];
+        for (const [index, field] of encryptedFields.fields.entries()) {
```

Comment:
I don't think this logic is correct.  The algorithm outlined in the spec says

- for each entry in the encrypted fields map
- if the keyId is null, create it.  if there's an error, return early with the partially filled encryptedFields.

We can't parallelize this, it should be sequential

Reply:
I had the same thought but came to the conclusion that this is probably okay – practically speaking, callers are still going to have to handle the same set of circumstances, right?

Reply:
I guess that's true.

Personally, I see it as an unnecessary deviation from the spec.  It's not a high-traffic code path, so I don't see parallelization being important.  Also

- the code is more complicated
- users need to contend with multiple errors and mapping errors to objects

If we do decide to go forward with this approach - should we consider storing more actionable information on the error than just an array of errors?  The current implementation doesn't provide an easy mechanism for users to map which error came from which null keyId.

Reply:
After hopping on a call with Bailey we've agreed that parallelism is okay, but we shouldn't try to have all this index tracking, it adding to the complexity of it all. I've updated things to use filter, map, and some where applicable, take another look.